### PR TITLE
Support "shared" test clusters

### PR DESCRIPTION
**Pull request by mark-vieira**
This commit introduces the concept of shared tests clusters using the new JUnit testing framework. Unlike normal test clusters which are exclusive to each test suite (class), shared clusters persist across test suites to be reused. This can be useful for test projects with a large number of test suites that can all use a single cluster, and the overhead of creating these clusters is the dominating factor.

### CODE REVIEW

**Comment by mark-vieira** [comment]
I've ported some of the EQL tests here which I thought were a good candidate because they have a relatively large number of suites in a single project. The multi-cluster tests in particular would be quite costly to run in a cluster-per-suite fashion.

**Comment by mark-vieira** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Comment by mark-vieira** [comment]
@elasticmachine update branch

File: test/test-clusters/src/main/java/org/elasticsearch/test/cluster/local/AbstractLocalClusterSpecBuilder.java [hunk 1]
```diff
@@ -107,12 +108,34 @@ public AbstractLocalClusterSpecBuilder<T> rolesFile(Resource rolesFile) {
         return this;
     }
 
+    @Override
+    public LocalClusterSpecBuilder<T> shared(Boolean isShared) {
+        if (Integer.parseInt(System.getProperty("tests.max.parallel.forks")) > 1) {
```
[the remark this answers is not in the document]
Comment (breskeby) [reply test/test-clusters/src/main/java/org/elasticsearch/test/cluster/local/AbstractLocalClusterSpecBuilder.java:113]: I just checked locally and the tradeoff is little for the win we get. Also we should nudge devs to run single tests when developping or thats what IDEA is doing anyhow during dev

**Review by breskeby — APPROVED** [review]
lgtm.

File: test/test-clusters/src/main/java/org/elasticsearch/test/cluster/local/AbstractLocalClusterSpecBuilder.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mark-vieira) [reply test/test-clusters/src/main/java/org/elasticsearch/test/cluster/local/AbstractLocalClusterSpecBuilder.java:113]: Yeah, the main reason to enforce this is that you cannot guarantee which tests run on which vm, and therefore which shared cluster. There may be cases in which that is problematic.

[the remark this answers is not in the document]
Comment (breskeby) [reply test/test-clusters/src/main/java/org/elasticsearch/test/cluster/local/AbstractLocalClusterSpecBuilder.java:113]: we could probably tweak the logic to have a cluster per forked jvm. But likely not something we need now.

[the remark this answers is not in the document]
Comment (mark-vieira) [reply test/test-clusters/src/main/java/org/elasticsearch/test/cluster/local/AbstractLocalClusterSpecBuilder.java:113]: Well, that's exactly what would happen. Which is _probably_ fine but I think in most cases it's more efficient to just force a single cluster.

**Pull request merged**